### DISCUSSION TITLE
Crm crashes after triggering custom workflow by modifying record

### DISCUSSION THREAD

**Discussion by kseibi** [YetiForceCompany/YetiForceCRM#14660] [How to] [answered]
Recently I have been working on a custom workflow that can be triggered on modification of record, I understand that I need to 

1. First, create a file with the function code and save it to modules/moudulename/workflows/, 
2. Register the workflow to com_vtiger_workflow_entitymethods table in the database
3. Choose "invoke custom function" in the "Add to do"

I imitated the functions in modules/HelpDesk/workflows/HelpDeskWorkflow.php and wrote the function that I need, uploaded it to modules/Vendors/workflows/

However for the second step, i couldn't register the function using these lines,

```
include_once 'include/main/WebUI.php';
require 'modules/com_vtiger_workflow/VTEntityMethodManager.inc';
$emm = new VTEntityMethodManager(PearDatabase::getInstance()); 
$emm->addEntityMethod('Vendors', 'myfunction', 'modules/Vendors/workflows/myfunctionclass.php', 'myfunctionclass');
```

When I run http://IP/registerworkflow.php the webpage says "crm is currently unable to handle this request.", and nothing happens in com_vtiger_workflow_entitymethods table, so I decided to add the method manually to com_vtiger_workflow_entitymethods, and it worked

I could use the function and it works fine when I choose "Manual trigger by the user", but **whenever I choose "Every time a record is saved (eg. creating, modifying)", and go on to modify a record, when I try to save it just keeps loading and the whole server crashes, I would have to restart the server to use CRM again,** 

I noticed that if I get rid of "$recordModel->save();" in my function, it would work and the server wouldn't crash, however, I need this to save values to the record fields

I tried to make the function as easy as modifying a field value and saving it, but still didn't work

This is my code:

```
class myfunctionclass
{
	public static function myfunction(Vtiger_Record_Model $recordModel)
	{
		\App\Log::trace('Entering myfunction');         
                $name= 'myname';
                $recordModel->set('name', $name);
		$recordModel->save();
		\App\Log::trace('myfunction');
	}
}

```

I need the workflow to trigger on record modification, what am i doing wrong??

**Comment by rskrzypczak** [comment, marked as answer]
Your code loops, so it's no wonder your system freezes.
check this:
https://github.com/YetiForceCompany/YetiForceCRM/blob/e0d22cea149f977acb275b603d97b264f7753d3e/modules/com_vtiger_workflow/tasks/VTUpdateFieldsTask.php#L29-L32
and
https://github.com/YetiForceCompany/YetiForceCRM/blob/e0d22cea149f977acb275b603d97b264f7753d3e/modules/com_vtiger_workflow/tasks/VTUpdateFieldsTask.php#L66

**Reply by kseibi** [reply]
I added these and changed the parameter of the function and it worked! Thanks for the help

**Comment by mariuszkrzaczkowski** [comment]
> ```
> include_once 'include/main/WebUI.php';
> require 'modules/com_vtiger_workflow/VTEntityMethodManager.inc';
> $emm = new VTEntityMethodManager(PearDatabase::getInstance()); 
> $emm->addEntityMethod('Vendors', 'myfunction', 'modules/Vendors/workflows/myfunctionclass.php', 'myfunctionclass');
> ```
> 
> When I run http://IP/registerworkflow.php the webpage says "crm is currently unable to handle this request.", and nothing happens in com_vtiger_workflow_entitymethods table, so I decided to add the method manually to com_vtiger_workflow_entitymethods, and it worked

it should be like this:

```php
require_once 'modules/com_vtiger_workflow/VTEntityMethodManager.php';
$emm = new VTEntityMethodManager();
$emm->addEntityMethod('Vendors', 'myfunction', 'modules/Vendors/workflows/myfunctionclass.php', 'myfunctionclass');
```

https://github.com/YetiForceCompany/YetiForceCRM/blob/e0d22cea149f977acb275b603d97b264f7753d3e/modules/Settings/Workflows/models/Module.php#L290-L292
